uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestStatusMatrix.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+use CRM_HRLeaveAndAbsences_Service_LeaveManager as LeaveManagerService;
+use CRM_HRLeaveAndAbsences_BAO_LeaveRequest as LeaveRequest;
+
+class CRM_HRLeaveAndAbsences_Service_LeaveRequestStatusMatrix {
+
+  /**
+   * @var \CRM_HRLeaveAndAbsences_Service_LeaveManager
+   */
+  private $leaveManagerService;
+
+  /**
+   * @var array|null
+   *   Stores the list of option values for the LeaveRequest status_id field.
+   */
+  private static $leaveStatuses;
+
+  /**
+   * CRM_HRLeaveAndAbsences_Service_LeaveRequestStatusMatrix constructor.
+   *
+   * @param \CRM_HRLeaveAndAbsences_Service_LeaveManager $leaveManagerService
+   */
+  public function __construct(LeaveManagerService $leaveManagerService) {
+    $this->leaveManagerService = $leaveManagerService;
+  }
+
+  /**
+   * Checks whether it is possible for the current user to change a leave request
+   * status from fromStatus to toStatus given the leave request contact id
+   * using the status matrix applicable to the current user
+   *
+   * @param int $fromStatus
+   * @param int $toStatus
+   * @param int $leaveRequestContactID
+   *
+   * @return bool
+   */
+  public function canTransitionTo($fromStatus, $toStatus, $leaveRequestContactID) {
+    $currentUserID = CRM_Core_Session::getLoggedInContactID();
+    $statusMatrix = [];
+
+    if ($currentUserID == $leaveRequestContactID){
+      $statusMatrix = $this->getStaffStatusMatrix();
+    }
+
+    if ($this->shouldUseManagerMatrixForCurrentUser($leaveRequestContactID)) {
+      $statusMatrix = $this->getManagerStatusMatrix();
+    }
+
+    if (empty($statusMatrix)) {
+      throw new RuntimeException('You are not authorized to update this leave request');
```

Comment:
I'm not sure about this exception here. I think that, even for cases like this, we should return false. The point is, this class is only concerned about the matrix and transitioning from one status to another, but the exception is saying that we cannot update a leave request, doesn't really have anything to do with status.